src/moonlink/src/table_handler.rs

```diff
@@ -31,6 +31,150 @@ pub struct EventSyncSender {
+    // - For non-streaming events, they start with a [`Begin`] message containing the final LSN of the current transaction, which we could leverage to decide keep or not.
+    initial_persistence_lsn: Option<u64>,
+    // Record LSN if the last handled table event is committed, which indicates mooncake table stays at a consistent view, so table could be flushed safely.
+    table_consistent_view_lsn: Option<u64>,
+    // Latest LSN of the table's latest commit.
+    latest_commit_lsn: Option<u64>,
+
+    // table management and event handling states
+    //
+    // Whether iceberg snapshot result has been consumed by the latest mooncake snapshot, when creating a mooncake snapshot.
+    //
+    // There're three possible states for an iceberg snapshot:
+    // - snapshot ongoing = false, result consumed = true: no active iceberg snapshot
+    // - snapshot ongoing = true, result consumed = true: iceberg snapshot is ongoing
+    // - snapshot ongoing = false, result consumed = false: iceberg snapshot completes, but wait for mooncake snapshot to consume the result
+    //
+    iceberg_snapshot_result_consumed: bool,
+    iceberg_snapshot_ongoing: bool,
+    // Whether there's an ongoing background maintenance operation, for example, index merge, data compaction, etc.
+    // To simplify state management, we have at most one ongoing maintainance operation at the same time.
+    maintainance_ongoing: bool,
+    // Whether there's an ongoing mooncake snapshot operation.
+    mooncake_snapshot_ongoing: bool,
+    // Pending force snapshot requests.
+    pending_force_snapshot_lsns: BTreeMap<u64, Vec<Option<Sender<Result<()>>>>>,
+    // Special table state, for example, initial copy, alter table, drop table, etc.
+    special_table_state: SpecialTableState,
+    // Buffered events during blocking operations: initial copy, alter table, drop table, etc.
+    initial_copy_buffered_events: Vec<TableEvent>,
+}
+
+impl TableHandlerState {
+    fn new() -> Self {
+        Self {
+            iceberg_snapshot_result_consumed: true,
+            iceberg_snapshot_ongoing: false,
+            maintainance_ongoing: false,
+            mooncake_snapshot_ongoing: false,
+            initial_persistence_lsn: None,
+            table_consistent_view_lsn: None,
+            latest_commit_lsn: None,
+            pending_force_snapshot_lsns: BTreeMap::new(),
+            special_table_state: SpecialTableState::Normal,
+            initial_copy_buffered_events: Vec::new(),
+        }
+    }
+
+    fn update_table_lsns(&mut self, event: &TableEvent) {
+        if event.is_ingest_event() {
+            match event {
+                TableEvent::Commit { lsn, .. } => {
+                    self.latest_commit_lsn = Some(*lsn);
+                    self.table_consistent_view_lsn = Some(*lsn);
+                }
+                _ => {
+                    self.table_consistent_view_lsn = None;
```

Comment:
table consistent view LSN is used to indicate whether we could safely flush, so for table maintenance events like force snapshot, index merge, etc; it should keep unchanged